### ISSUE TITLE
Add vector regridding to new mediator

### ISSUE THREAD

**Issue by rmontuoro** [SWPC-IPE/WAM-IPE#78]
Code to handle regridding of vector fields (@oehmke), available in the current mediator, needs to be added to the recently introduced new mediator (see issue #70). We will follow the steps below:

- [ ] Add code to new mediator in branch updated_mediator (https://github.com/SWPC-IPE/WAM-IPE/tree/updated_mediator)

- [ ] Run coupled simulations and compare results with old mediator

- [ ] Debug if needed

- [ ] Perform code review

- [ ] Merge into development after current updated_mediator branch

**Comment by rmontuoro** [comment]
The [vector_regridding_validate](https://github.com/SWPC-IPE/WAM-IPE/tree/vector_regridding_validate) branch was created off of [vector_regridding](https://github.com/SWPC-IPE/WAM-IPE/tree/vector_regridding) feature branch at commit 69f1a8e58a2ffb065b5a3037a50a5c0891705d90 to include debug code for additional field output.

This branch can be deleted once validation is complete and all relevant commits are moved to original [vector_regridding](https://github.com/SWPC-IPE/WAM-IPE/tree/vector_regridding) branch.

**Comment by rmontuoro** [comment]
The sources of the observed differences in the early 5-day regression test for branch [vector_regridding](https://github.com/SWPC-IPE/WAM-IPE/tree/vector_regridding) were traced back to the the following items:

1. Conversion of imported `height` field from _m_ to _km_ was performed by 
- dividing by `1000.0` in the Spaceweather mediator:
https://github.com/SWPC-IPE/WAM-IPE/blob/d9bc6ed62e7a467e082a98400acdfbb955063855/NEMS/src/module_MEDIATOR_SpaceWeather.F90#L1618-L1619
- multiplying by `1._ESMF_KIND_R8/1000.0` while in the SWPC mediator:
https://github.com/SWPC-IPE/WAM-IPE/blob/d9bc6ed62e7a467e082a98400acdfbb955063855/NEMS/src/module_MEDIATOR_SWPC.F90#L466-L467
These different conversion strategies caused small numerical differences that were amplified in successive calculations. Restoring the previous conversion method in the SWPC mediator (commit 5b1dca14deff84b7b8e81512fce0054117fe3c8c in [vector_regridding_validate](https://github.com/SWPC-IPE/WAM-IPE/tree/vector_regridding_validate)) eliminated the observed differences for all fields except `N2_Density`, which showed smaller relative differences than before (of the order of `5.E-16`).

2. In the Spaceweather mediator, the molecular weight of gaseous species is used independently from the temperature values during vertical extrapolation. For example, see:
https://github.com/SWPC-IPE/WAM-IPE/blob/d9bc6ed62e7a467e082a98400acdfbb955063855/NEMS/src/module_MEDIATOR_SpaceWeather.F90#L1718
In the SWPC mediator, the temperature is normalized by the molecular weight before being passed to the vertical interpolation/extrapolation routine:
https://github.com/SWPC-IPE/WAM-IPE/blob/d9bc6ed62e7a467e082a98400acdfbb955063855/NEMS/src/module_MEDIATOR_SWPC_methods.F90#L2995
https://github.com/SWPC-IPE/WAM-IPE/blob/d9bc6ed62e7a467e082a98400acdfbb955063855/NEMS/src/module_MEDIATOR_SWPC_methods.F90#L3000-L3002
Using temperature and mass independently in the SWPC mediator eliminated completely all discrepancies (commit 5b1dca14deff84b7b8e81512fce0054117fe3c8c in [vector_regridding_validate](https://github.com/SWPC-IPE/WAM-IPE/tree/vector_regridding_validate)).

Twin 2-day runs using `march2013_regression.config` with `FHMAX=2` and `FHRES=2` generated _bitwise identical_ results between the Spaceweather and SWPC mediator when ESMF settings where used to ensure bitwise reproducibility of regridding results in different runs (commit 8fefa72bb8d0ef3858a60278b26c8be7dc2cd630). The output examined included additional single and double precision data from each WAM field, saved at each coupling time step (40 total) at different stages in the mediator's execution flow (_e.g._: pre- and post-regridding, in IPE, in IPELIB, etc.).

A similar 5-day twin run (`FHMAX=120`, `FHRES=120`) also showed _bitwise identical_ results between the Spaceweather and SWPC mediator.

**Comment by rmontuoro** [comment]
Commits 5b1dca14deff84b7b8e81512fce0054117fe3c8c and 8fefa72bb8d0ef3858a60278b26c8be7dc2cd630 in the [vector_regridding_validate](https://github.com/SWPC-IPE/WAM-IPE/tree/vector_regridding_validate) branch were moved to the original [vector_regridding](https://github.com/SWPC-IPE/WAM-IPE/tree/vector_regridding) branch as follows:

- Commit 5b1dca14deff84b7b8e81512fce0054117fe3c8c was ported as two separate commits:
15f3d329b9a8ff99fdb78b5f68d58dd8c92889e3: 
Added normalization option to `NamespaceSetRemoteLevelsFromField()`. This option is now used when converting heights from m to km in the SWPC mediator to replicate the behavior of the Spaceweather mediator.
4edbba57164f5f8e2d289987f1e97cbcbfe50aa3:
Provided molecular weight of neutral gases as argument in `LogInterpolate()` to be used independentlty of temperature during vertical interpolation/extrapolation in the SWPC mediator. This replicates the implementation in the Spaceweather mediator.

- Commit 8fefa72bb8d0ef3858a60278b26c8be7dc2cd630 became 7a115c5a0a69e130e05e0b9d4db317a9edc42abd

Again, 2-day and 5-day twin regression tests performed with the [vector_regridding](https://github.com/SWPC-IPE/WAM-IPE/tree/vector_regridding) branch at 7a115c5a0a69e130e05e0b9d4db317a9edc42abd returned _bitwise identical_ results.

**Comment by rmontuoro** [comment]
Branch [development](https://github.com/SWPC-IPE/WAM-IPE/tree/development) at f9cccf2 was merged into feature branch [vector_regridding](https://github.com/SWPC-IPE/WAM-IPE/tree/vector_regridding) at 7a115c5 (commit e3d2062c2c78f3566e38c8739fa34601f8513f3e). 

ESMF settings to ensure bitwise reproducibility of regridding results across different runs are left _on_ by default, and may be turned off when bringing this branch into [development](https://github.com/SWPC-IPE/WAM-IPE/tree/development) by commenting out the following line in `NEMS/src/makefile`:
https://github.com/SWPC-IPE/WAM-IPE/blob/e3d2062c2c78f3566e38c8739fa34601f8513f3e/NEMS/src/makefile#L275

The [development](https://github.com/SWPC-IPE/WAM-IPE/tree/development) branch introduced an updated `march2013_regression.config` regression test that uses a `58x8` IPE decomposition _vs._ the original `1x80`.

Twin 2-hour runs using the new regression test with `FHMAX=2`,`FHRES=2`, and both Spaceweather and SWPC mediators returned _bitwise identical_ results. 
Twin 5-day runs with `FHMAX=120` and `FHRES=120` did not complete within the assigned 8-hour wallclock time on Theia since the code was considerably slower than before the merger. This might be due to the different IPE PET decomposition (`58x8`), or to additional features introduced in the [development](https://github.com/SWPC-IPE/WAM-IPE/tree/development) branch. However, all of the 21 output files generated with the Spaceweather and SWPC mediators were _bitwise identical_.

**Comment by rmontuoro** [comment]
Issued pull request #158.
